### PR TITLE
add more params and functionalities taking advantage of GraphicsMagick

### PR DESCRIPTION
**Pull request by phamhongviet**
The added params are:
* gravity: "-gravity" param, default is Center
* crop: "-crop" param with "+repage"
* rotate: "-rotate" param
* monochrome: "-monochrome" param
* grey: "-colorspace" param with "GRAY"
* no_strip: "-strip" param
* trim: "-trim" param
* no_interlace: "-interlace" param with "Line". Useful for JPEG, I'm not sure about other formats.
* flip: "-flip" param
* flop: "-flop" param
Please go to GraphicsMagick documentation for detailed explanation.

Added test functions for buildArguments.

Params parser now divided into 3 main parts to parse int, bool and string type params. Adding params is easier this way.

### CODE REVIEW

**Comment by pierrre** [comment]
tests failed: https://travis-ci.org/pierrre/imageserver/jobs/93622418#L412-L415

File: graphicsmagick/graphicsmagick.go [hunk 1]
```diff
@@ -29,23 +30,44 @@ const (
...
+// - no_strip: "-strip" param
+//
+// - trim: "-trim" param
+//
+// - no_interlace: "-interlace" param with "Line"
+//
+// - flip: "-flip" param
+//
+// - flop: "-flop" param
+
```
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: https://travis-ci.org/pierrre/imageserver/jobs/93622418#L760-L762
Remove this new line.

**Comment by phamhongviet** [comment]
Well this is embarrassing. May I ask what version of GraphicsMagick you are using? Mine is GraphicsMagick 1.3.21 2015-02-28 Q8.

**Comment by pierrre** [comment]
Debian uses GraphicsMagick 1.3.12...
https://travis-ci.org/pierrre/imageserver/jobs/93639075#L111

**Comment by pierrre** [comment]
Do you need the `-strip` argument?
You could skip the test if the GraphicsMagick version is too old.

File: graphicsmagick/graphicsmagick.go [hunk 2]
```diff
@@ -29,23 +30,41 @@ const (
 //
 // Params:
 //
-// - width / height: sizes for "-resize" argument (both optionals)
```
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: Why did you move this line?
I think `width` and `height` are the most important parameters.

File: graphicsmagick/graphicsmagick.go [hunk 3]
```diff
@@ -314,6 +377,134 @@ func (server *Server) buildArgumentsQuality(arguments *list.List, params imagese
 	return nil
 }
 
+func (server *Server) buildArgumentsGravity(arguments *list.List, params imageserver.Params) error {
+	gravity, _ := params.GetString("gravity")
```
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: You must check the error.
Call `params.Has()` before.

File: graphicsmagick/graphicsmagick.go [hunk 4]
```diff
@@ -314,6 +377,134 @@ func (server *Server) buildArgumentsQuality(arguments *list.List, params imagese
...
+	var translatedGravity string
+	if gravity != "" {
```
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: remove this condition

File: graphicsmagick/graphicsmagick.go [hunk 5]
```diff
@@ -314,6 +377,134 @@ func (server *Server) buildArgumentsQuality(arguments *list.List, params imagese
...
+			translatedGravity = "NorthEast"
+		case gravity == "se":
+			translatedGravity = "SouthEast"
+		case gravity == "nw":
+			translatedGravity = "NorthWest"
+		case gravity == "sw":
+			translatedGravity = "SouthWest"
+		}
+		if translatedGravity == "" {
+			return &imageserver.ParamError{Param: "gravity", Message: "gravity should n, s, e, w, ne, se, nw or sw"}
```
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: It's the `default` in the `switch/case`

File: graphicsmagick/graphicsmagick.go (hunk 4 above)
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: `if params.Has("gravity")`

File: graphicsmagick/graphicsmagick.go [hunk 6]
```diff
@@ -29,23 +30,41 @@ const (
...
+// - quality: "-quality" param
+//
+// - background: color for "-background" argument, 3/4/6/8 lower case hexadecimal characters
 //
 // - format: "-format" param
 //
-// - quality: "-quality" param
+// - gravity: "-gravity" param, default is Center
+//
+// - crop: "-crop" param with "+repage"
```
Comment (pierrre) [inline graphicsmagick/graphicsmagick.go]: I don't understand.
You can already "crop" wtih `extent=true`+`fill=true`.

**Comment by pierrre** [comment]
I'm sorry, but with your pull request, a simple URL like `http://localhost:8080/?source=large.jpg&width=100&height=100` doesn't work anymore...
The image is not resized. (I agree, the current code is not well tested)
Did you run your code?

I run the current code in production and it works well.
You add new options, and I'm not really confident with their behaviors:
- `-crop` (you can already use `extent=true`+`fill=true`)
- `-gravity` is always added to the command line even if it's not used (by extent or crop)
- `no_interlace` looks weird. What is the default value used by GraphicsMagick? What's the behavior if the image is not JPEG? Should we expose more options (plane, partition...)? If not, shouldn't we call this parameter `progressive` instead of `no_interlace`?

Do you really need theses new parameters?
Will you run this code in production?
I think you could copy the file `graphicsmagick.go`, customize it, and paste it to your project.
https://github.com/pierrre/imageserver is a toolkit and it's easy to add your own components.

By the way, I think that this "GraphicsMagick server" is awkward:
- it uses an external program, and you can't bundle with your Go program (and there are issues with available features across versions)
- it starts a new process for each image
- Go forks the current (Go) process if you start an external process
- a fork is slow if the current process uses more than 100MB of memory
- you can't use "in memory" cache or groupcache, because they increase the memory usage of the Go process

You should try my new integrations with GIFT or nfnt/resize.
https://godoc.org/github.com/pierrre/imageserver/image/gift
It's pure Go, and doesn't start an external process.
It's also easier to compose image transformations, because you don't need to write a monolithic component like GraphicsMagick.

**Comment by pierrre** [comment]
Actually, I wanted to deprecate the GraphicsMagick `Server` and remove it.
But I need to keep it, since someone could still use it.

**Comment by phamhongviet** [comment]
Hi Pierre,

`strip` remove the text attributes of an image to make it lighter and more suitable for web and mobile apps.

I didn't remove width and height, but I did mess with the order by putting it behind extent. Sorry about that.

About `crop`, I have never combining "extent" and "fill" before. I didn't know that you can crop like this. I think adding crop is useful since the param name is what it really does and you can crop a specific area
using x and y offset. This is the crop documentation from GraphicsMagick website: http://www.graphicsmagick.org/GraphicsMagick.html#details-crop

About `no_interlace`, it's useful for JPEG but I heard it hurts PNG compression. So may be we can check for the format before apply this or just remove it.

I agree with you on using GraphicsMagick. I've seen a 4GB server dedicated to running GraphicsMagick run out of memory. And writing an image to disk to be manipulated by GraphicsMagick and read it again from disk destroy the performance unless we mount the temp dir with tmpfs or ramfs.

I do use all the added parameters in production. I've already created a forked version of your graphicsmagick.go in my project so it's really doesn't affect me much if this pull request is not merged. In stead of width and height, I have to use w and h. That's why I messed up with the param order in the comment :-p.

**Comment by phamhongviet** [comment]
Using GIFT instead of GraphicsMagick looks very promising! I was trying to docker containerize my image server with Alpine base. I end up using CentOS base, which is a LOT bigger than Alpine, because GraphicsMagick isn't available in Alpine haha. With GIFT, my docker image could be much smaller.
Please drop this pull request if it's broken for you. I'm using my fork for my production code anyway.

**Comment by pierrre** [comment]
> strip remove the text attributes of an image to make it lighter and more suitable for web and mobile apps.

Yes I know.
I you use the GIFT (or nfnt/resize) processor, it will always strip EXIF metadata.
Go image package doesn't support EXIF.

> I didn't remove width and height, but I did mess with the order by putting it behind extent. Sorry about that.

Sorry, my `graphicsmagick` was not 100% unit tested and it didn't detect the error.
But my new `gift` package is 100% tested: https://codecov.io/github/pierrre/imageserver

> About crop, I have never combining "extent" and "fill" before. I didn't know that you can crop like this. I think adding crop is useful since the param name is what it really does and you can crop a specific area
> using x and y offset. This is the crop documentation from GraphicsMagick website: http://www.graphicsmagick.org/GraphicsMagick.html#details-crop

I already know the `-crop` option.
But I think `fill + extent` is more powerful: you can use `fill` without `extent`.
And I don't know how it works if you combine fill+extent+crop.  What is the expected behavior? Will it crash? Is it an error?

> About no_interlace, it's useful for JPEG but I heard it hurts PNG compression. So may be we can check for the format before apply this or just remove it.

By the way, the Go `image/jpeg` package doesn't support progressive encoding.
https://groups.google.com/forum/#!searchin/golang-nuts/jpeg/golang-nuts/6HZKs4YCnmM/LCjSBS9sCQAJ
If you use my `gift` package, you can't choose the encoding method.

> I agree with you on using GraphicsMagick. I've seen a 4GB server dedicated to running GraphicsMagick run out of memory. And writing an image to disk to be manipulated by GraphicsMagick and read it again from disk destroy the performance unless we mount the temp dir with tmpfs or ramfs.

You can use: https://godoc.org/github.com/pierrre/imageserver#NewLimitServer
It limits the concurrent executions (and the memory usage).

> I do use all the added parameters in production. I've already created a forked version of your graphicsmagick.go in my project so it's really doesn't affect me much if this pull request is not merged. In stead of width and height, I have to use w and h. That's why I messed up with the param order in the comment :-p.

If you want to change the name of the parameters in the URL, you don't need to change `graphicsmagick/graphicsmagick.go`.
You just need to edit the parser in `http/graphicsmagick/graphicsmagick.go` (and create a copy).
You can keep the same internal names for the parameters.

**By the way, you must always check and handle the errors.
I spotted some mistakes in this PR.**

> Using GIFT instead of GraphicsMagick looks very promising!

It's easier to combine image filter/transformation (image transformation and encoding are separated).
It allows me to do gamma correction: https://github.com/pierrre/imageserver#gamma-correction-more-info
I plan to use it in production soon.

> Please drop this pull request if it's broken for you. I'm using my fork for my production code anyway.

OK :(
I'm not really confident with some changes.
If you can use your own copy of `graphicsmagick.go`, it's fine :)
